File: frontend/iosApp/UseCases/Sources/UseCases/DomainMappers/MemoryMapper.swift
import Foundation
import Domains
import APIGateways
import Utilities

public enum MemoryMapper {
    /// Map from server response (generates new localId)
    public static func toDomain(_ response: MemoryResponse, albumLocalId: UUID) -> Memory? {
        let imageUrl: URL? = {
            if let uri = response.imageLocalUri {
                return URL(string: uri)
            }
            if let url = response.imageRemoteUrl {
                return URL(string: url)
            }
            return nil
        }()

        guard let imageUrl else { return nil }

        return Memory(
            serverId: response.id,
            localId: UUID(),
            albumId: response.albumId,
            albumLocalId: albumLocalId,
            title: response.title,
            imageUrl: imageUrl,
            imageLocalPath: nil,
            createdAt: DateFormatters.iso8601.date(from: response.createdAt) ?? Date(),
            syncStatus: .synced
        )
    }

    /// Map from server response (preserves existing localId)
    public static func toDomain(_ response: MemoryResponse, localId: UUID, albumLocalId: UUID) -> Memory? {
        let imageUrl: URL? = {
            if let uri = response.imageLocalUri {
                return URL(string: uri)
            }
            if let url = response.imageRemoteUrl {
                return URL(string: url)
            }
            return nil
        }()

        guard let imageUrl else { return nil }

        return Memory(
            serverId: response.id,
            localId: localId,
            albumId: response.albumId,
            albumLocalId: albumLocalId,
            title: response.title,
            imageUrl: imageUrl,
            imageLocalPath: nil,
            createdAt: DateFormatters.iso8601.date(from: response.createdAt) ?? Date(),
            syncStatus: .synced
        )
    }
}

File: frontend/shared/src/commonMain/kotlin/com/example/memoriesapp/mapper/MemoryMapper.kt
package com.example.memoriesapp.mapper

import com.example.memoriesapp.api.response.MemoryResponse
import com.example.memoriesapp.core.LocalId
import com.example.memoriesapp.core.SyncStatus
import com.example.memoriesapp.core.Timestamp
import com.example.memoriesapp.domain.Memory

/**
 * Mapper for Memory domain model
 */
object MemoryMapper {
    /**
     * Convert API response to domain model (generates new localId)
     */
    fun toDomain(response: MemoryResponse, albumLocalId: LocalId): Memory? {
        val imageUrl = response.imageLocalUri ?: response.imageRemoteUrl ?: return null

        return Memory(
            serverId = response.id,
            localId = LocalId.generate(),
            albumId = response.albumId,
            albumLocalId = albumLocalId,
            title = response.title,
            imageUrl = imageUrl,
            imageLocalPath = null,
            createdAt = parseTimestamp(response.createdAt),
            syncStatus = SyncStatus.SYNCED
        )
    }

    /**
     * Convert API response to domain model (preserves existing localId)
     */
    fun toDomain(response: MemoryResponse, localId: LocalId, albumLocalId: LocalId): Memory? {
        val imageUrl = response.imageLocalUri ?: response.imageRemoteUrl ?: return null

        return Memory(
            serverId = response.id,
            localId = localId,
            albumId = response.albumId,
            albumLocalId = albumLocalId,
            title = response.title,
            imageUrl = imageUrl,
            imageLocalPath = null,
            createdAt = parseTimestamp(response.createdAt),
            syncStatus = SyncStatus.SYNCED
        )
    }

    private fun parseTimestamp(dateString: String): Timestamp {
        return try {
            Timestamp.parse(dateString)
        } catch (e: Exception) {
            Timestamp.now()
        }
    }
}
